src/engine/Math/graph.ts

```diff
@@ -0,0 +1,809 @@
+        }
+
+        // Find the edge connecting current to neighbor
+        const edge: Edge<T> = Array.from(currentNode.edges).find((e: Edge<T>) => e.source.id === currentId && e.target.id === neighborId);
+
+        if (!edge) {
+          continue;
+        }
+
+        cameFrom.set(neighborId, currentId);
+        // Add to open set if not already there
+        if (!openSet.has(neighborId)) {
+          openSet.add(neighborId);
+        }
+      }
+    }
+
+    // No path found
+    return { path: [], pathSteps: 0, distance: Infinity, skippedNodes };
+  }
+
+  private _euclideanDistance(currentNode: PositionNode<T>, testNode: PositionNode<T>): number {
+    const a = currentNode.pos;
+    const b = testNode.pos;
+    return Math.sqrt((b.x - a.x) ** 2 + (b.y - a.y) ** 2);
+  }
+}
+
+/**
+ * Represents an edge in a graph, connecting two nodes.
+ * @template T The type of data stored in the nodes connected by this edge.
+ */
+export class Edge<T> {
+  private _id: G_UUID = GraphUUId.generateUUID();
+  private _source: Node<T>;
+  private _target: Node<T>;
+  private _weight: number = 0;
+  private _partnerEdge: Edge<T> | null = null; // Reference to the opposite direction edge
+
+  constructor(source: Node<T>, target: Node<T>, weight: number = 0) {
+    this._source = source;
+    this._target = target;
+    this._weight = weight;
+  }
+
+  linkWithPartner(partnerEdge: Edge<T>): void {
+    this._partnerEdge = partnerEdge;
+  }
+
+  get id() {
+    return this._id;
+  }
+
+  get source() {
+    return this._source;
+  }
+
+  get target() {
+    return this._target;
+  }
+
+  get weight() {
+    return this._weight;
+  }
+
+  get partnerEdge() {
+    return this._partnerEdge;
+  }
+}
+
+/**
+ * Represents a node in a graph, with a unique identifier and optional data.
+ * @template T The type of data stored in this node.
+ */
+export class Node<T> {
+  private _id: G_UUID = GraphUUId.generateUUID();
+  private _data: T;
+  private _edges: Set<Edge<T>>;
+
+  constructor(data: T) {
+    this._data = data;
+    this._edges = new Set();
+  }
+
+  get id(): G_UUID {
+    return this._id;
+  }
+
+  get data() {
+    return this._data;
+  }
+
+  get edges() {
+    return this._edges;
+  }
+
+  registerNewEdge(newEdge: Edge<T>) {
+    this._edges.add(newEdge);
+  }
+
+  breakEdge(edge: Edge<T>) {
+    this._edges.delete(edge);
+  }
+
+  getConnectedNodes(): Node<T>[] {
+    return Array.from(this._edges).map((edge) => edge.target);
+  }
+}
+
+/**
+ * Represents a node in a graph with a unique identifier, optional data, and a position in space.
+ * @template T The type of data stored in this node.
+ * @augments {Node<T>}
+ */
+export class PositionNode<T> extends Node<T> {
+  pos: Vector;
+
+  constructor(data: T, pos: Vector) {
+    super(data);
+    this.pos = pos;
+  }
+}
+
+/**
+ * Alias for a node in a graph, representing a vertex in a geometric context.
+ * @template T The type of data stored in this vertex.
+ */
+export type Vertex<T> = Node<T>;
```

Comment:
Should we stick to either node or vertex? I'm a little worried the synonym could be confusing.

Reply:
It seems there are duplicate methods with different names

Reply:
I would actually prefer to keep it to just nodes, and remove vertex